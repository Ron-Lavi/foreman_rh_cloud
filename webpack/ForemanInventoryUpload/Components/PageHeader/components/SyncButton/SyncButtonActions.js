import React from 'react';
import { get, post } from 'foremanReact/redux/API';
import { withInterval } from 'foremanReact/redux/middlewares/IntervalMiddleware';
import { addToast } from 'foremanReact/redux/actions/toasts';
import { translate as __ } from 'foremanReact/common/I18n';
import { inventoryUrl } from '../../../../ForemanInventoryHelpers';
import Toast from './components/Toast';
import {
  INVENTORY_SYNC,
  INVENTORY_SYNC_TASK_UPDATE,
} from './SyncButtonConstants';
import { foremanUrl } from '../../../../../ForemanRhCloudHelpers';

export const handleSync = () => dispatch => {
  dispatch(
    post({
      key: INVENTORY_SYNC,
      url: inventoryUrl('tasks'),
      handleSuccess: ({
        data: {
          task: { id },
        },
      }) => {
        dispatch(getSyncTaskInterval(id));
        return dispatch(
          taskPageRefererToast(id, 'info', __('Inventory sync has started:'))
        );
      },
      errorToast: inventorySyncErrorToast,
    })
  );
};

export const getSyncTaskInterval = id => dispatch => {
  dispatch(
    withInterval(
      get({
        key: INVENTORY_SYNC_TASK_UPDATE,
        url: inventoryUrl(`tasks/${id}`),
        handleSuccess: ({ data: { result, output } }, stopTaskInterval) => {
          if (result === 'success') {
            const {
              host_statuses: { sync, disconnect },
            } = output;
            dispatch(
              addToast({
                sticky: true,
                type: 'success',
                message: (
                  <Toast syncHosts={sync} disconnectHosts={disconnect} />
                ),
              })
            );
          }
          if (result === 'error') {
            dispatch(
              taskPageRefererToast(
                id,
                'error',
                __('Inventory sync has failed:'),
                true
              )
            );
          }
          stopTaskInterval();
        },
        errorToast: inventorySyncErrorToast,
      })
    )
  );
};

const inventorySyncErrorToast = ({ message, response }) =>
  `${__('Inventory sync has failed: ')} ${response.data?.message || message}`;

const taskPageRefererToast = (taskID, toastType, prefix, sticky = false) =>
  addToast({
    sticky,
    type: toastType,
    message: (
      <span>
        {prefix}{' '}
        <a
          target="_blank"
          rel="noopener noreferrer"
          href={foremanUrl(`/foreman_tasks/tasks/${taskID}`)}
        >
          {__('view the task page for more details')}
        </a>
      </span>
    ),
  });
