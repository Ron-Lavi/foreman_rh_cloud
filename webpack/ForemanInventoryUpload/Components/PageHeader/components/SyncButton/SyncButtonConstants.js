export const INVENTORY_SYNC = 'INVENTORY_SYNC';

export const INVENTORY_SYNC_TASK_UPDATE = 'INVENTORY_SYNC_TASK_UPDATE';
