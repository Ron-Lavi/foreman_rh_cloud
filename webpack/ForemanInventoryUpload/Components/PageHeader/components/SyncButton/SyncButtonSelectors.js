import { selectAPIResponse } from 'foremanReact/redux/API/APISelectors';
import { INVENTORY_SYNC_TASK_UPDATE } from './SyncButtonConstants';

export const selectTaskStatus = state => {
  const { result } = selectAPIResponse(state, INVENTORY_SYNC_TASK_UPDATE);
  return typeof result === 'string' ? result.toUpperCase() : null;
};
