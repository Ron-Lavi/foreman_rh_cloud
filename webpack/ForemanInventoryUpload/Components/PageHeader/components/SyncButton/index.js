import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';

import * as actions from './SyncButtonActions';
import SyncButton from './SyncButton';
import { selectCloudToken } from '../../../InventorySettings/InventorySettingsSelectors';
import { selectTaskStatus } from './SyncButtonSelectors';

// map state to props
const mapStateToProps = state => ({
  cloudToken: selectCloudToken(state),
  status: selectTaskStatus(state),
});

// map action dispatchers to props
const mapDispatchToProps = dispatch => bindActionCreators(actions, dispatch);

// export connected component
export default connect(mapStateToProps, mapDispatchToProps)(SyncButton);
