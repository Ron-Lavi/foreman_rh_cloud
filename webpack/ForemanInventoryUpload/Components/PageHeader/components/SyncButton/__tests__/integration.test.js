import React from 'react';
import axios from 'axios';
import { IntegrationTestHelper, MockAdapter } from '@theforeman/test';
import SyncButton from '../index';
import { successResponse } from './SyncButtonFixtures';
import apiReducer from '../../../../../../__mocks__/foremanReact/redux/API/APIReducer';
import { APIMiddleware } from '../../../../../../__mocks__/foremanReact/redux/API/APIMiddleware';

const mock = new MockAdapter(axios);
mock.onPost().reply(({ url }) => new Promise(resolve => successResponse));

describe('SyncButton integration test', () => {
  it('should flow', async () => {
    const integrationTestHelper = new IntegrationTestHelper(
      { API: apiReducer },
      [APIMiddleware]
    );
    const wrapper = integrationTestHelper.mount(<SyncButton />);
    const instance = wrapper.find('SyncButton').instance();
    instance.props.handleSync();
    await IntegrationTestHelper.flushAllPromises();
    integrationTestHelper.takeStoreAndLastActionSnapshot(
      'handleSync was called'
    );
  });
});
