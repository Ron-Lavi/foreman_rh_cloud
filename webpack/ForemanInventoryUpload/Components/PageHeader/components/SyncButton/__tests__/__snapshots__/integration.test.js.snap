// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SyncButton integration test should flow: handleSync was called 1`] = `
Object {
  "action": Array [
    Array [
      Object {
        "payload": Object {
          "errorToast": [Function],
          "handleSuccess": [Function],
          "key": "INVENTORY_SYNC",
          "url": "/foreman_inventory_upload/tasks",
        },
        "type": "API_POST",
      },
    ],
  ],
  "state": Object {
    "API": Object {
      "INVENTORY_SYNC": Object {
        "payload": Object {
          "url": "/foreman_inventory_upload/tasks",
        },
        "response": null,
        "status": "PENDING",
      },
    },
  },
}
`;
